File: pkg/contract/taskflow.go
// SPDX-License-Identifier: AGPL-3.0-only

package contract

type TaskFlow struct {
	SchemaVersion string            `json:"schemaVersion"`
	ID            string            `json:"id"`
	Name          string            `json:"name"`
	Game          GameKey           `json:"game"`
	Servers       []ServerKey       `json:"servers"`
	Resolutions   []Resolution      `json:"resolutions"`
	Entrypoint    string            `json:"entrypoint"`
	Tasks         []TaskDefinition  `json:"tasks"`
	Metadata      map[string]string `json:"metadata,omitempty"`
}

type TaskDefinition struct {
	ID        TaskID            `json:"id"`
	Name      string            `json:"name"`
	Steps     []TaskStep        `json:"steps"`
	OnFailure FailurePolicy     `json:"onFailure"`
	Produces  []string          `json:"produces,omitempty"`
	Metadata  map[string]string `json:"metadata,omitempty"`
}

type TaskStep struct {
	ID          string         `json:"id"`
	Description string         `json:"description,omitempty"`
	Primitive   string         `json:"primitive"`
	Params      map[string]any `json:"params,omitempty"`
	When        string         `json:"when,omitempty"`
	Next        string         `json:"next,omitempty"`
	OnFailure   FailurePolicy  `json:"onFailure,omitempty"`
	TimeoutMS   int            `json:"timeoutMs,omitempty"`
}

type FailurePolicy struct {
	Severity     Severity `json:"severity"`
	RetryLimit   int      `json:"retryLimit,omitempty"`
	RetryDelayMS int      `json:"retryDelayMs,omitempty"`
	FallbackStep string   `json:"fallbackStep,omitempty"`
}

File: contracts/rust/actingcommand-contract-rs/src/taskflow.rs
// SPDX-License-Identifier: AGPL-3.0-only

//! Rust backup data structures for declarative task-flow contracts.

use crate::types::*;
use std::collections::BTreeMap;

#[derive(Debug, Clone, PartialEq)]
pub struct TaskFlow {
    pub schema_version: String,
    pub id: String,
    pub name: String,
    pub game: GameKey,
    pub servers: Vec<ServerKey>,
    pub resolutions: Vec<Resolution>,
    pub entrypoint: String,
    pub tasks: Vec<TaskDefinition>,
    pub metadata: Metadata,
}

#[derive(Debug, Clone, PartialEq)]
pub struct TaskDefinition {
    pub id: TaskId,
    pub name: String,
    pub steps: Vec<TaskStep>,
    pub on_failure: FailurePolicy,
    pub produces: Vec<String>,
    pub metadata: Metadata,
}

#[derive(Debug, Clone, PartialEq)]
pub struct TaskStep {
    pub id: String,
    pub description: Option<String>,
    pub primitive: String,
    pub params: BTreeMap<String, TaskParamValue>,
    pub when: Option<String>,
    pub next: Option<String>,
    pub on_failure: Option<FailurePolicy>,
    pub timeout_ms: Option<DurationMillis>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct FailurePolicy {
    pub severity: Severity,
    pub retry_limit: Option<i32>,
    pub retry_delay_ms: Option<DurationMillis>,
    pub fallback_step: Option<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum TaskParamValue {
    Null,
    Bool(bool),
    Number(f64),
    String(String),
    Array(Vec<TaskParamValue>),
    Object(BTreeMap<String, TaskParamValue>),
}
